src/Foundatio.Aliyun/Storage/AliyunFileStorage.cs

```diff
@@ -0,0 +1,199 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Aliyun.OSS;
+using Foundatio.Extensions;
+
+namespace Foundatio.Storage {
+    public class AliyunFileStorage : IFileStorage {
+        private readonly string _bucketName;
+        private readonly OssClient _client;
+
+        public AliyunFileStorage(string connectionString, string bucketName = "storage") {
+            var account = AliyunConnectionString.Parse(connectionString);
+            _client = account.CreateClient();
+            _bucketName = bucketName;
+            if (!_client.DoesBucketExist(_bucketName)) _client.CreateBucket(_bucketName);
+        }
+
+        public void Dispose() { }
+
+        private bool IsNotFoundException(Exception ex) {
+            if (ex is AggregateException aggregateException) {
+                foreach (var innerException in aggregateException.InnerExceptions) {
+                    if (IsNotFoundException(innerException)) {
+                        return true;
+                    }
+                }
+            }
+            if (ex is WebException webException && webException.Response is HttpWebResponse response) {
+                return response.StatusCode == HttpStatusCode.NotFound;
+            }
+            return false;
+        }
+
+        private string NormalizePath(string path) {
+            return path?.Replace('\\', '/');
+        }
+
+        public async Task<Stream> GetFileStreamAsync(string path, CancellationToken cancellationToken = default(CancellationToken)) {
+            if (string.IsNullOrEmpty(path)) {
+                throw new ArgumentNullException(nameof(path));
+            }
+            var response = await Task.Factory.FromAsync(

```

Comment:
A part of me thinks that if they ossclient doesn't support async maybe we should just use their sync methods until the time they update it? Is this how you are calling it in your own projects (in an async manner)? I Tried searching their issues but they are not in English. I'm really surprised they don't have an async api.

Reply:
The OssClient only provided BeginXXX and EndXXX methods to implement asynchronous pattern. After investigation, the AzureStorage SDK also implemented the async method with the same approach to encapsulate BeginXXX and EndXXX methods. Please refer to https://github.com/Azure/azure-storage-net/blob/bdeb107c5fc1ad6db84fe6deaff17b15717aac0c/Lib/ClassLibraryCommon/Blob/CloudBlobContainer.cs#L155